src/hotspot/share/utilities/bitMap.hpp

```diff
@@ -354,84 +315,125 @@ class BitMap {
+    // Remove previous bits - no need to clear
+    resize(0, false /* clear */);
+
+    initialize(new_size_in_bits, clear);
+  }
+
+  // Protected functions, that are used by BitMap sub-classes that support them.
+
+  // Resize the backing bitmap memory.
+  //
+  // Old bits are transferred to the new memory
+  // and the extended memory is cleared.
+  void resize(idx_t new_size_in_bits, bool clear = true) {
+    bm_word_t* new_map = reallocate(_allocator, map(), size(), new_size_in_bits, clear);
+    update(new_map, new_size_in_bits);
+  }
+};
+
 // A concrete implementation of the "abstract" BitMap class.
 //
 // The BitMapView is used when the backing storage is managed externally.
 class BitMapView : public BitMap {
  public:
-  BitMapView() : BitMap(NULL, 0) {}
+  BitMapView() : BitMapView(nullptr, 0) {}
   BitMapView(bm_word_t* map, idx_t size_in_bits) : BitMap(map, size_in_bits) {}
 };
 
-// A BitMap with storage in a ResourceArea.
-class ResourceBitMap : public BitMap {
+class CHeapBitMapAllocator : StackObj {
+  MEMFLAGS _flags;
 
  public:
-  ResourceBitMap() : BitMap(NULL, 0) {}
-  // Conditionally clears the bitmap memory.
-  ResourceBitMap(idx_t size_in_bits, bool clear = true);
+  CHeapBitMapAllocator(MEMFLAGS flags) : _flags(flags) {}
+  BitMap::bm_word_t* allocate(size_t size_in_words) const;
 
-  // Resize the backing bitmap memory.
-  //
-  // Old bits are transferred to the new memory
-  // and the extended memory is cleared.
-  void resize(idx_t new_size_in_bits);
+  void free(BitMap::bm_word_t* map, BitMap::idx_t size_in_words) const;
+};
 
-  // Set up and clear the bitmap memory.
-  //
-  // Precondition: The bitmap was default constructed and has
-  // not yet had memory allocated via resize or initialize.
-  void initialize(idx_t size_in_bits);
+class ArenaBitMapAllocator : StackObj {
+  Arena* const _arena;
 
-  // Set up and clear the bitmap memory.
-  //
-  // Can be called on previously initialized bitmaps.
-  void reinitialize(idx_t size_in_bits);
+ public:
+  ArenaBitMapAllocator(Arena* arena) : _arena(arena) {}
+  BitMap::bm_word_t* allocate(BitMap::idx_t size_in_words) const;
+  void free(BitMap::bm_word_t* map, BitMap::idx_t size_in_words) const {
+    // ArenaBitMaps currently don't free memory.
+  }
 };
 
 // A BitMap with storage in a specific Arena.
-class ArenaBitMap : public BitMap {
+class ArenaBitMap : public GrowableBitMap<ArenaBitMapAllocator> {
  public:
   // Clears the bitmap memory.
-  ArenaBitMap(Arena* arena, idx_t size_in_bits);
+  ArenaBitMap(Arena* arena, idx_t size_in_bits, bool clean = true);
 
- private:
-  NONCOPYABLE(ArenaBitMap);
+ protected:
+  ArenaBitMap(const ArenaBitMap& rhs) = default;
+  ArenaBitMap& operator=(const ArenaBitMap& rhs) = delete;
 };
 
-// A BitMap with storage in the CHeap.
-class CHeapBitMap : public BitMap {
+// A BitMap with storage in a ResourceArea. It is an ArenaBitMap but _arena is nullptr;
+class ResourceBitMap : public ArenaBitMap {
```

Comment:
How do we want to treat resource allocations which look like arena allocations? The compiler code does have a places where objects are arena allocated in `Thread::current()->resource_area()`. I am not sure about having this behaviour in the type system. 

The places where `VectorSet()` is used today, can we not just use `ArenaBitMap(Thread::current()->resource_area(), [...])` instead. And have the replacement be `VectorSet` -> `ArenaBitMap`

Reply:
This is how I treat them using the base pointer 'ArenaBitMap*'
https://github.com/navyxliu/jdk/commit/cdbc0e1c97f747fd53ad408d59e65ec8adf0fc16#diff-17a9d48a5570e5bcde557c8d1efcbe78d69a5657bc865bed37b22c3c8b5d447d

> The places where VectorSet() is used today, can we not just use ArenaBitMap(Thread::current()->resource_area(), [...]) instead. And have the replacement be VectorSet -> ArenaBitMap

`Thread::current()->resource_area()` returns ResourceArena*. ResourceArena is a subclass of Arena. In this sense, we can use the same ArenaAllocator to handle both ResourceBitMap and ArenaBitMap. That's why I remove 'ResourceAllocator' in this patch. 

I think it is possible to do what you said. Let's say we use ArenaBitMap as 'ResourceBitMap'. Shall we keep ArenaBitMap and ResourceBitMap separated then? In classic object-oriented design, ResourceBitMap 'is-a' ArenaBitMap, isn't it? If we keep 2 classes in parallel, we may end up 2 'ResourceBitMap' with different behaviors. 

In current patch, ResourceBitMap is just an adapter. It supports copy constructor and copy assignment because 'resource arena' is unique.

Reply:
Correction: s/resource arena/resource area

This is my understanding. 'Resource area' in HotSpot is a special 'area' whose lifecycle is bound to stacktrace. By default, we still use 'arena' allocation algorithm. That's why I think we can treat ResourceBitMap as a subclass of ArenaBitMap.

Reply:
I like the type hierarchy but for one thing. Today we have two ways of creating a resource arena
* `ResourceBitmap([...])`
  * Which checks `verify_has_resource_mark` in debug builds
* `ArenaBitmap(some_resource_area, [...])`
  * Which does not check `verify_has_resource_mark`

This would add a third way where `ArenaBitmap(nullptr, [...])` has the same behaviour as `ResourceBitmap`. This can be solved with a new `ArenaBitMap` constructor.
```c++
 protected:
  ArenaBitMap(std::nullptr_t, idx_t size_in_bits, bool clean);
```
And add a debug assert in the normal ArenaBitMap constructor `precond(arena != nullptr);`

Which would give a behaviour like this
```c++
#ifdef ASSERT
TEST_VM(BitMap, resource_work) {
  ResourceBitMap rbm(0);
  ArenaBitMap abm(Thread::current()->resource_area(),0);
  static_cast<void>(rbm);
  static_cast<void>(abm);
}
TEST_VM_ASSERT(BitMap, nullptr_arena) {
  Arena* ptr = nullptr;
  ArenaBitMap crash(ptr,0);
  // ArenaBitMap does_not_compile(nullptr,0);
  // ArenaBitMap does_not_compile(NULL,0);
  static_cast<void>(crash);
}
#endif
```

I do not know if the extra branch in the arena allocator ever matters, but if it did you can just store the resource area arena in `_arena` on release and not branch. The only semantical difference then would be that if the [re]allocating functions gets called from different threads. But I would think that a resource allocated bitmap crossing thread boundaries is never an appropriate use. 

But I did preface this discussion with that I do not know how we treat resource allocated objects that just look like arena allocated objects and thus circumvent the `verify_has_resource_mark` check. It happens all over the place in the compiler code, and I am not sure why.

Reply:
First of all, I revert back ResourceBitMap. It's now independent to ArenaBitMap. I think you concern is reasonable. In most cases, I expect users to use ResourceBitMap. ArenaBitmap does accept ResourceArea* as the underlying arena because Unique_Node_List uses it in this way. 

I don't think we need to handle ArenaBitMap::_arena == nullptr like you said. ResourceArea is still an Arena, but is managed by ResourceMark automatically. As a result, I introduce a boolean flag `Arena::_mark_managed`. ArenaBitMap doesn't mess with the underlying storage if Arena::mark_managed() is true. 

Here is my experimental patch.
https://github.com/navyxliu/jdk/commit/88063064a3bc0b8e65d3834f13e5ea8c6576eab0#diff-17a9d48a5570e5bcde557c8d1efcbe78d69a5657bc865bed37b22c3c8b5d447d

We have to support copy assignment for Unique_Node_List. It won't introduce memory leak because we only use ArenaBitMap in certain scenarios, see the assert. Besides ResourceArena, Unique_Node_List only accepts Compile::comp_arena() || Compiler::node_arena(). The lifecycle of them are same as Compile object. once a compilation unit is done, both two arenas are released completely.